File: src/main/java/org/taktik/icure/entities/embed/Content.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.entities.embed;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.squareup.moshi.Json;
import org.jetbrains.annotations.Nullable;
import org.taktik.icure.utils.InstantDeserializer;
import org.taktik.icure.utils.InstantSerializer;

import java.io.Serializable;
import java.time.Instant;
import java.util.List;
import java.util.Map;


@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Content implements Serializable {
	@JsonProperty("s")
	@Json(name = "s")
	String stringValue;

	@JsonProperty("n")
	@Json(name = "n")
	Double numberValue;

	@JsonProperty("b")
	@Json(name = "b")
	Boolean booleanValue;

	@JsonProperty("i")
	@Json(name = "i")
	@JsonSerialize(using = InstantSerializer.class, include = JsonSerialize.Inclusion.NON_NULL)
	@JsonDeserialize(using = InstantDeserializer.class)
	Instant instantValue;

	@JsonProperty("dt")
	@Json(name = "dt")
	Long fuzzyDateValue;

	@JsonProperty("x")
	@Json(name = "x")
	byte[] binaryValue;

	@JsonProperty("d")
	@Json(name = "d")
	String documentId;

	@JsonProperty("m")
	@Json(name = "m")
	Measure measureValue;

	@JsonProperty("p")
	@Json(name = "p")
	Medication medicationValue;

	@JsonProperty("c")
    @Json(name = "c")
    List<Service> compoundValue;

	@JsonIgnore
	byte[] compressedStringValue;

    public Content() {
    }

    public Content(String stringValue) {
        this.stringValue = stringValue;
    }

    public Content(Double numberValue) {
        this.numberValue = numberValue;
    }

    public Content(Boolean booleanValue) {
        this.booleanValue = booleanValue;
    }

    public Content(Instant instantValue) {
        this.instantValue = instantValue;
    }

    public Content(Measure measureValue) {
        this.measureValue = measureValue;
    }

    public Content(byte[] binaryValue) {
        this.binaryValue = binaryValue;
    }

    public Content(Medication medicationValue) {
        this.medicationValue = medicationValue;
    }

    public @Nullable String getStringValue() {
        return stringValue;
    }

    public void setStringValue(String stringValue) {
        this.stringValue = stringValue;
    }

    public @Nullable Double getNumberValue() {
        return numberValue;
    }

    public void setNumberValue(Double numberValue) {
        this.numberValue = numberValue;
    }

    public @Nullable Boolean getBooleanValue() {
        return booleanValue;
    }

    public void setBooleanValue(Boolean booleanValue) {
        this.booleanValue = booleanValue;
    }

    public @Nullable Instant getInstantValue() {
        return instantValue;
    }

    public void setInstantValue(Instant instantValue) {
        this.instantValue = instantValue;
    }

	public  @Nullable Long getFuzzyDateValue() {
		return fuzzyDateValue;
	}

	public void setFuzzyDateValue(Long fuzzyDateValue) {
		this.fuzzyDateValue = fuzzyDateValue;
	}

	public byte[] getBinaryValue() {
        return binaryValue;
    }

    public void setBinaryValue(byte[] binaryValue) {
        this.binaryValue = binaryValue;
    }

    public @Nullable String getDocumentId() {
        return documentId;
    }

    public void setDocumentId(String documentId) {
        this.documentId = documentId;
    }

    public @Nullable Measure getMeasureValue() {
        return measureValue;
    }

    public void setMeasureValue(Measure measureValue) {
        this.measureValue = measureValue;
    }

    public @Nullable Medication getMedicationValue() {
        return medicationValue;
    }

    public void setMedicationValue(Medication medicationValue) {
        this.medicationValue = medicationValue;
    }

	public List<Service> getCompoundValue() {
		return compoundValue;
	}

	public void setCompoundValue(List<Service> compoundValue) {
		this.compoundValue = compoundValue;
	}

	public byte[] getCompressedStringValue() {
		return compressedStringValue;
	}

	public void setCompressedStringValue(byte[] compressedStringValue) {
		this.compressedStringValue = compressedStringValue;
	}
}

File: src/main/kotlin/org/taktik/icure/entities/embed/Content.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.entities.embed

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.github.pozo.KotlinBuilder
import org.taktik.icure.utils.InstantDeserializer
import org.taktik.icure.utils.InstantSerializer
import java.io.Serializable
import java.time.Instant

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class Content(
        @JsonProperty("s") val stringValue: String? = null,
        @JsonProperty("n") val numberValue: Double? = null,
        @JsonProperty("b") val booleanValue: Boolean? = null,
        @JsonProperty("i")
        @JsonSerialize(using = InstantSerializer::class)
        @JsonInclude(JsonInclude.Include.NON_NULL)
        @JsonDeserialize(using = InstantDeserializer::class)
        val instantValue: Instant? = null,
        @JsonProperty("dt") val fuzzyDateValue: Long? = null,
        @JsonProperty("x") val binaryValue: ByteArray? = null,
        @JsonProperty("d") val documentId: String? = null,
        @JsonProperty("m") val measureValue: Measure? = null,
        @JsonProperty("p") val medicationValue: Medication? = null,
        @JsonProperty("c") val compoundValue: Set<Service>? = null,
        val ratio: List<Measure>? = null,
        val range: List<Measure>? = null
) : Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Content) return false

        if (stringValue != other.stringValue) return false
        if (numberValue != other.numberValue) return false
        if (booleanValue != other.booleanValue) return false
        if (instantValue != other.instantValue) return false
        if (fuzzyDateValue != other.fuzzyDateValue) return false
        if (binaryValue != null) {
            if (other.binaryValue == null) return false
            if (!binaryValue.contentEquals(other.binaryValue)) return false
        } else if (other.binaryValue != null) return false
        if (documentId != other.documentId) return false
        if (measureValue != other.measureValue) return false
        if (medicationValue != other.medicationValue) return false
        if (compoundValue != other.compoundValue) return false
        if (ratio != other.ratio) return false
        if (range != other.range) return false

        return true
    }

    override fun hashCode(): Int {
        var result = stringValue?.hashCode() ?: 0
        result = 31 * result + (numberValue?.hashCode() ?: 0)
        result = 31 * result + (booleanValue?.hashCode() ?: 0)
        result = 31 * result + (instantValue?.hashCode() ?: 0)
        result = 31 * result + (fuzzyDateValue?.hashCode() ?: 0)
        result = 31 * result + (binaryValue?.contentHashCode() ?: 0)
        result = 31 * result + (documentId?.hashCode() ?: 0)
        result = 31 * result + (measureValue?.hashCode() ?: 0)
        result = 31 * result + (medicationValue?.hashCode() ?: 0)
        result = 31 * result + (compoundValue?.hashCode() ?: 0)
        result = 31 * result + (ratio?.hashCode() ?: 0)
        result = 31 * result + (range?.hashCode() ?: 0)
        return result
    }
}
